add "app.kubernetes.io/name" to gateway.podlabels

lukmdo opened the pull request:
**Please provide a description of this PR:**

Two commits to squash or keep the _main_ only

- main (separate commit)
> add "app.kubernetes.io/name" to `gateway.podLabels`
>
> Goal: to have matching "app.kubernetes.io/name" label and value for on `deployment` and `pod` for metrics.
>
>Using `.Values.labels` for that will lead to duplicate "app.kubernetes.io/name" labels.

- https://helm.sh/docs/chart_best_practices/labels/#standard-labels
- https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels

---
- optional (cleanup as separate commit)

> move `istio.io/rev` label to `gateway.podLabels` (not-functional change)

lukmdo committed:
add "app.kubernetes.io/name" to `gateway.podLabels`

Goal: to have matching "app.kubernetes.io/name" label and value for on `deployment` and `pod` for metrics.

Using `.Values.labels` for that will lead to duplicate "app.kubernetes.io/name" labels.

lukmdo committed:
opt: (cleanup) move `istio.io/rev` label to `gateway.podLabels`

this is not-functional change only optional cleanup

istio-testing commented:
Hi @lukmdo. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

hanxiaop reviewed (commented):
`app` label can take effect too for the telemetry, see https://istio.io/latest/docs/ops/deployment/requirements/#pod-requirements

costinm commented on manifests/charts/gateway/templates/_helpers.tpl:
```diff
@@ -30,6 +30,10 @@ app.kubernetes.io/name: {{ include "gateway.name" . }}
 {{ include "gateway.selectorLabels" . }}
+app.kubernetes.io/name: {{ include "gateway.name" . }}
+{{- with .Values.revision }}
+istio.io/rev: {{ . | quote }}
```
Should we also add app.kubernetes.io/version	(or use that instead of istio.io/rev ) ? 

Also would be nice if we added the K8S recommended labels consistently to all our templates - consistency is important.

lukmdo replied:
added in 281c532ee5 

and optional cleanup in 3ac9a50475

marking as resolved - please reopen anytime with new comments 

+1 consistency

costinm reviewed (approved):
Looks good, useful for telemetry - but I would use the standard label for version too.

lukmdo committed:
review: add app.kubernetes.io/version to gateway.podLabels

https://github.com/istio/istio/pull/50540#pullrequestreview-2011501679

lukmdo committed:
review: (opt cleanup) use `gateway.podLabels` in `gateway.labels`

https://github.com/istio/istio/pull/50540#pullrequestreview-2011501679

help managing common labels

lukmdo commented:
> Looks good, useful for telemetry - but I would use the standard label for version too.

thank you for feedback 

@costinm I added separate new commits 
- 281c532ee5 review: add `app.kubernetes.io/version` to `gateway.podLabels`
- 3ac9a50475  review: (opt cleanup) use `gateway.podLabels` in `gateway.labels`

ericvn commented:
I'll add a DNM which @costinm can remove after he reviews the additional changes.

costinm commented on manifests/charts/gateway/templates/_helpers.tpl:
```diff
@@ -15,21 +15,19 @@ Create chart name and version as used by the chart label.
-{{ $key | quote }}: {{ $val | quote }}
-{{- end }}
-{{- end }}
+{{ include "gateway.podLabels" . }}
```
I am quite confused by this refactoring. Why would the gateway labels include pod labels ? Can we keep the code clean - even if it means some duplication of lines, there is no reason for all other resources to have the injection labels ( labels are indexed in k8s, so more expensive - and also create noise ). It's already pretty messy due to the 'reuse' and lack of clear spec on what labels we support for each object type - selectorLabels is also specific to pods, no need to put them on service or other objects.

My concern here is that we'll accidentally add even more labels or break backwards compat, and the code is already hard to read/review. If is intended for all resources - should be in gateway.labels, if it is pod specific - in podLabels or directly on the pod.

lukmdo replied:
please see https://github.com/istio/istio/pull/50540/files#r1582266135

- `master` currently has 
```
gateway.labels
  - gateway.selectorLabels

gateway.podLabels
  - gateway.selectorLabels

gateway.selectorLabels
```

my initial idea was to reuse the common labels between `gateway.labels` and `gateway.podLabels` by moving them to `gateway.podLabels`
```
gateway.labels
  - gateway.podLabels

gateway.podLabels
  - gateway.selectorLabels

gateway.selectorLabels
```

The problem with that and the current `gateway.podLabels` is that it represents "gateway.commonLabels"

Please see https://github.com/istio/istio/pull/50540/files#r1582266135 which has 
```
gateway.labels
  - gateway.selectorLabels

gateway.selectorLabels

gateway.sidecarInjectionLabels
```

costinm reviewed (commented):
Left few comments on readability/structure.

lukmdo committed:
review: add gateway.sidecarInjectionLabels, drop gateway.podLabels

lukmdo commented on manifests/charts/gateway/templates/deployment.yaml:
```diff
@@ -24,9 +24,6 @@ spec:
-        {{- with .Values.revision }}
-        istio.io/rev: {{ . | quote }}
-        {{- end }}
         {{- include "gateway.podLabels" . | nindent 8 }}
```
`gateway.podLabels`  is only used here and in `gateway.labels`

I agree naming here may be misleading as the current `gateway.podLabels` is more: 
- `gateway.podBaseLabels`
- `gateway.commonLabels`

(no injection labels)

I decided to apply your comments in 5f32127f2c0ffc85ccd21d5cbff1cd91ac8ad85a
 by: 
- add `sidecarInjectionLabels` with note and doc link
- remove the `gateway.podLabels` template.  If it feels easier to read and maintain, may consider using `gateway.commonLabels`
```
{{- define "gateway.commonLabels" -}}
{{ include "gateway.selectorLabels" }}
app.kubernetes.io/name: {{ include "gateway.name" . }}
app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
{{- range $key, $val := .Values.labels }}
{{- if and (ne $key "app") (ne $key "istio") }}
{{ $key | quote }}: {{ $val | quote }}
{{- end }}
{{- end }}
{{- end }}
```

lukmdo commented on manifests/charts/gateway/templates/deployment.yaml:
```diff
@@ -24,9 +24,6 @@ spec:
       {{- end }}
       labels:
         sidecar.istio.io/inject: "true"
```
Thank you - applied by addding `sidecarInjectionLabels` with note and doc link

please see comment https://github.com/istio/istio/pull/50540/files#r1582266135

lukmdo commented:
See: 
<details>
  <summary><b>Helm Diff: master/pr50540</b></summary>

```diff
--- helm.master.yaml	2024-04-28 18:57:22
+++ helm.pr50540.yaml	2024-04-28 19:10:10
@@ -1,188 +1,190 @@
 ---
 # Source: gateway/templates/serviceaccount.yaml
 apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: istio-gateway
   namespace: remote-ag-stg
   labels:
     helm.sh/chart: gateway-1.0.0
     app: istio-gateway
     istio: gateway
+    app.kubernetes.io/name: istio-gateway
     app.kubernetes.io/version: "1.0.0"
     app.kubernetes.io/managed-by: Helm
-    app.kubernetes.io/name: istio-gateway
 ---
 # Source: gateway/templates/role.yaml
 apiVersion: rbac.authorization.k8s.io/v1
 kind: Role
 metadata:
   name: istio-gateway
   namespace: remote-ag-stg
   labels:
     helm.sh/chart: gateway-1.0.0
     app: istio-gateway
     istio: gateway
+    app.kubernetes.io/name: istio-gateway
     app.kubernetes.io/version: "1.0.0"
     app.kubernetes.io/managed-by: Helm
-    app.kubernetes.io/name: istio-gateway
   annotations:
     {}
 rules:
 - apiGroups: [""]
   resources: ["secrets"]
   verbs: ["get", "watch", "list"]
 ---
 # Source: gateway/templates/role.yaml
 apiVersion: rbac.authorization.k8s.io/v1
 kind: RoleBinding
 metadata:
   name: istio-gateway
   namespace: remote-ag-stg
   labels:
     helm.sh/chart: gateway-1.0.0
     app: istio-gateway
     istio: gateway
+    app.kubernetes.io/name: istio-gateway
     app.kubernetes.io/version: "1.0.0"
     app.kubernetes.io/managed-by: Helm
-    app.kubernetes.io/name: istio-gateway
   annotations:
     {}
 roleRef:
   apiGroup: rbac.authorization.k8s.io
   kind: Role
   name: istio-gateway
 subjects:
 - kind: ServiceAccount
   name: istio-gateway
 ---
 # Source: gateway/templates/service.yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: istio-gateway
   namespace: remote-ag-stg
   labels:
     helm.sh/chart: gateway-1.0.0
     app: istio-gateway
     istio: gateway
+    app.kubernetes.io/name: istio-gateway
     app.kubernetes.io/version: "1.0.0"
     app.kubernetes.io/managed-by: Helm
-    app.kubernetes.io/name: istio-gateway
   annotations:
     {}
 spec:
   type: LoadBalancer
   ports:
     - name: status-port
       port: 15021
       protocol: TCP
       targetPort: 15021
     - name: http2
       port: 80
       protocol: TCP
       targetPort: 80
     - name: https
       port: 443
       protocol: TCP
       targetPort: 443
   selector:
     app: istio-gateway
     istio: gateway
 ---
 # Source: gateway/templates/deployment.yaml
 apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: istio-gateway
   namespace: remote-ag-stg
   labels:
     helm.sh/chart: gateway-1.0.0
     app: istio-gateway
     istio: gateway
+    app.kubernetes.io/name: istio-gateway
     app.kubernetes.io/version: "1.0.0"
     app.kubernetes.io/managed-by: Helm
-    app.kubernetes.io/name: istio-gateway
   annotations:
     {}
 spec:
   selector:
     matchLabels:
       app: istio-gateway
       istio: gateway
   template:
     metadata:
       annotations:
         inject.istio.io/templates: gateway
         prometheus.io/path: /stats/prometheus
         prometheus.io/port: "15020"
         prometheus.io/scrape: "true"
         sidecar.istio.io/inject: "true"
       labels:
         sidecar.istio.io/inject: "true"
         app: istio-gateway
         istio: gateway
+        app.kubernetes.io/name: istio-gateway
+        app.kubernetes.io/version: "1.0.0"
     spec:
       serviceAccountName: istio-gateway
       securityContext:
         # Safe since 1.22: https://github.com/kubernetes/kubernetes/pull/103326
         sysctls:
         - name: net.ipv4.ip_unprivileged_port_start
           value: "0"
       containers:
         - name: istio-proxy
           # "auto" will be populated at runtime by the mutating webhook. See https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#customizing-injection
           image: auto
           securityContext:
             capabilities:
               drop:
               - ALL
             allowPrivilegeEscalation: false
             privileged: false
             readOnlyRootFilesystem: true
             runAsUser: 1337
             runAsGroup: 1337
             runAsNonRoot: true
           env:
           ports:
           - containerPort: 15090
             protocol: TCP
             name: http-envoy-prom
           resources:
             limits:
               cpu: 2000m
               memory: 1024Mi
             requests:
               cpu: 100m
               memory: 128Mi
       terminationGracePeriodSeconds: 30
 ---
 # Source: gateway/templates/hpa.yaml
 apiVersion: autoscaling/v2
 kind: HorizontalPodAutoscaler
 metadata:
   name: istio-gateway
   namespace: remote-ag-stg
   labels:
     helm.sh/chart: gateway-1.0.0
     app: istio-gateway
     istio: gateway
+    app.kubernetes.io/name: istio-gateway
     app.kubernetes.io/version: "1.0.0"
     app.kubernetes.io/managed-by: Helm
-    app.kubernetes.io/name: istio-gateway
   annotations:
     {}
 spec:
   scaleTargetRef:
     apiVersion: apps/v1
     kind: Deployment
     name: istio-gateway
   minReplicas: 1
   maxReplicas: 5
   metrics:
     - type: Resource
       resource:
         name: cpu
         target:
           averageUtilization: 80
           type: Utilization
```
</details>

lukmdo commented:
> Left few comments on readability/structure.

@costinm thank you and I belive I applied/replied to all. 

I left some comments intentionally open for you to review. 
Please let me know if anything is outstanding that you would like to be done differently.
